lib/worker.js

```diff
@@ -330,12 +345,47 @@ class AssetComputeWorker {
         }
     }
 
+    sendEventsBeforeActionTimeout() {
+        return setTimeout(
+            async () => {
+                console.log(`Action is about to timeout in ${timeUntilTimeout()}ms. Sending Adobe I/O events before timeout.`);
+                const errorMessage = `Worker timed out without result after ${this.timers.actionDuration.end()} seconds`;
+                // ensure failure events are sent for any non successful rendition before timeout
+                if(this.renditions){
+                    for (const rendition of this.renditions) {
+                        if (!rendition.eventSent) {
+                            await this.events.sendEvent(EVENT_RENDITION_FAILED, {
+                                rendition: rendition.instructionsForEvent(),
+                                errorReason: Reason.GenericError,
+                                errorMessage: errorMessage,
+                                activationIds: this.params.customWorker ? [process.env.__OW_ACTIVATION_ID] : undefined
+                            });
+                            rendition.eventSent = true;
```

Comment:
I think in this case we should have the `rendition.eventSent = true;` _before_ the `await this.events.sendEvent()` to ensure another concurrently finishing processing does not send the event.

Reply:
Ah nice catch, I had actually run into this edge case in manual testing. I had changed it but I guess I forgot to commit my changes.

Reply:
I also found one more edge case in testing:

The `setTimeout` is defined in the [constructor](https://github.com/adobe/asset-compute-sdk/blob/ee89b351816af55d7640ebfae9d21df304fbd2ae/lib/worker.js#L77). And the send IO events, it relies on looking at the [`this.renditions`](https://github.com/adobe/asset-compute-sdk/blob/ee89b351816af55d7640ebfae9d21df304fbd2ae/lib/worker.js#L392) (which are all `Rendition` objects).  However, the `this.renditons` does not get defined until the [`prepare()`](https://github.com/adobe/asset-compute-sdk/blob/ee89b351816af55d7640ebfae9d21df304fbd2ae/lib/worker.js#L164) function, so if for some reason the code gets stuck and approaches time out before the `this.renditions` gets defined,  no events get sent. 

To handle this case, I went with the assumption that if `this.renditions` is not defined, we can assume no renditions have been successfully produced yet. This is because the code either errored or approached time out before it reached the point in the code where it defines the `this.renditions`. So, in this case, we can look at the `this.params.renditions` and send `rendition_failed` events for all of those renditions.

(btw I reached this edge case in testing by making a worker with a timeout = 15s so it would begin timing out immediately upon invoking the `worker()` function. I don't expect this case to come up too much in practice, but we should still handle it.)